### PR TITLE
add validating webhook

### PR DESCRIPTION
**Pull request by zhiweiyin318**

### CODE REVIEW

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+	clusterSetLabel = "cluster.open-cluster-management.io/clusterset"
+)
+
+func (a *AdmissionHandler) validateResource(request *v1.AdmissionRequest) *v1.AdmissionResponse {
+	status := &v1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	switch request.Resource {
+	case managedClustersGVR:
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:117]: I think you can use 
```go
case managedClustersGVR, clusterDeploymentsGVR, clusterClaimsGVR....
```

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	switch request.Resource {
+	case clusterDeploymentsGVR:
+		// ignore clusterDeployment created by clusterClaim
+		claimName, _, err := unstructured.NestedString(obj.Object, "spec", "clusterPoolRef", "claimName")
+		if err != nil {
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:166]: what if clusterpollref is not set? I think we should use type to decode here?

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+		} else if err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+				Message: err.Error(),
+			}
+			return status
+		}
+		return a.validateCreateRequest(request)
+	case v1.Update:
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:139]: this also need a.allowWithoutClusterSetLabel(request.UserInfo)

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+			Message: fmt.Sprintf("user %q is forbidden to create a resource without ManagedClusterSet label", request.UserInfo.Username),
+		}
+		return status
+	}
+
+	return a.allowSetClusterSetLabel(request.UserInfo, "", clusterSetName)
+}
+
+func (a *AdmissionHandler) validateUpdateRequest(request *v1.AdmissionRequest) *v1.AdmissionResponse {
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:197]: more doc here please.

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+		}
+		return a.validateCreateRequest(request)
+	case v1.Update:
+		return a.validateUpdateRequest(request)
+	}
+
+	return status
+}
+
+func (a *AdmissionHandler) validateCreateRequest(request *v1.AdmissionRequest) *v1.AdmissionResponse {
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:146]: more doc here please

File: pkg/webhook/clusterset/validatingWebhook.go [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+	if currentClusterSetName == "" {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: metav1.StatusReasonForbidden,
+			Message: fmt.Sprintf("user %q is forbidden to update a resource without ManagedClusterSet label", request.UserInfo.Username),
+		}
+		return status
+	}
+
+	return a.allowSetClusterSetLabel(request.UserInfo, originalClusterSetName, currentClusterSetName)
```
Comment (qiujian16) [inline pkg/webhook/clusterset/validatingWebhook.go:249]: To make it simple I think you only need a.allowSetClusterSetLabel(request.UserInfo, originalClusterSetName),

so if user did not set  clustersetname, you can still use this to func to check. For update operation, run the func twice with old and new clusterset  name.

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/webhook/clusterset/validatingWebhook.go]: done. I also move the codes handling `clusterdeployment` created by `clusterclaim`  here.

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 2 above)
Comment (zhiweiyin318) [reply pkg/webhook/clusterset/validatingWebhook.go]: done

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 3 above)
Comment (zhiweiyin318) [reply pkg/webhook/clusterset/validatingWebhook.go:157]: done. move a.allowWithoutClusterSetLabel(request.UserInfo) before Create and Update.

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 5 above)
Comment (zhiweiyin318) [reply pkg/webhook/clusterset/validatingWebhook.go:167]: done.
// validateCreateRequest validates:
// 1. the resource created must has a managedClusterSet label.
// 2. the user has managedClusterSet/join permission to create resource with a managedClusterSet label.

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 4 above)
Comment (zhiweiyin318) [reply pkg/webhook/clusterset/validatingWebhook.go:206]: done
// validateUpdateRequest validates:
// 1. allow if the managedClusterSet label is not changed.
// 2. the user must have clusterSet-A and clusterSet-B managedClusterSet/join permission if the managedClusterSet label is update from clusterSet-A to clusterSet-A.
// 3. the user must have clusterSet-A managedClusterSet/join permission if the managedClusterSet label clusterSet-A is removed.

File: pkg/webhook/clusterset/validatingWebhook.go (hunk 6 above)
Comment (zhiweiyin318) [reply pkg/webhook/clusterset/validatingWebhook.go]: done

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm
/hold

Feel free to unhold when you  think it is  fine to merge.

**Pull request merged**